mhs-reference-implementation/mhs/SDSLookup/sds.py

```diff
@@ -0,0 +1,105 @@
+import ssl
+from pathlib import Path
+
+from definitions import ROOT_DIR
+import ldap3
+
+from tornado.ioloop import IOLoop
+
+NHS_SERVICES_BASE = "ou=services, o=nhs"
+
+MHS_OBJECT_CLASS = "nhsMhs"
+AS_OBJECT_CLASS = "nhsAs"
+
+# These need replacing as the appropriate TLS certs are received for PTL
+PRIVATE_KEY = Path(ROOT_DIR) / 'data' / 'certs' / 'client.key'
+LOCAL_CERT_FILE = Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+CA_CERTS_FILE = Path(ROOT_DIR) / 'data' / 'certs' / 'client.cert'
+
+
+class SDSClient:
+    mhs_attributes = [
+        'nhsEPInteractionType', 'nhsIDCode', 'nhsMhsCPAId', 'nhsMHSEndPoint', 'nhsMhsFQDN',
+        'nhsMHsIN', 'nhsMHSIsAuthenticated', 'nhsMHSPartyKey', 'nhsMHsSN', 'nhsMhsSvcIA', 'nhsProductKey',
+        'uniqueIdentifier', 'nhsMHSAckRequested', 'nhsMHSActor', 'nhsMHSDuplicateElimination',
+        'nhsMHSPersistDuration', 'nhsMHSRetries', 'nhsMHSRetryInterval', 'nhsMHSSyncReplyMode'
+    ]
+
+    def __init__(self, ldap_address: str = None, tls: bool = False):
+        if ldap_address:
+            if tls:
+                load_tls = ldap3.Tls(local_private_key_file=PRIVATE_KEY,
+                                     local_certificate_file=LOCAL_CERT_FILE,
+                                     validate=ssl.CERT_REQUIRED, version=ssl.PROTOCOL_TLSv1,
+                                     ca_certs_file=CA_CERTS_FILE)
+
+                self.server = ldap3.Server(ldap_address, use_ssl=True, tls=load_tls)
+            else:
+                self.server = ldap3.Server(ldap_address)
+
+            self.connection = ldap3.Connection(self.server, auto_bind=True, client_strategy=ldap3.REUSABLE)
+        else:
+            self.server = None
+            self.connection = None
+
+    async def get_mhs_details(self, nhs_id_code: str, service_id: str):
+        """
+        Returns the all end point details for the given org code and service id
+        :param nhs_id_code:
+        :param service_id:
+        :return:
+        """
+
+        accredited_system_lookup = await self._accredited_system_lookup(nhs_id_code, service_id)
+
+        if not accredited_system_lookup:
+            raise Exception('No response from accredited system lookup')
```

Comment:
I added under the assumption it might be possible for a bad ods code/interaction id pair and that we would want to throw an error rather than the exception occurring when we try to index `None` for the party key.
I'll add a custom routing exception for the time being.